pilot/pkg/model/push_context.go

```diff
@@ -1657,43 +1685,182 @@ func (ps *PushContext) initSidecarScopes(env *Environment) error {
+	ps.sidecarIndex.sidecarsByNamespaceWithoutSelector = oldIndex.sidecarsByNamespaceWithoutSelector
+
+	for changedSidecar := range sidecarsToUpdate {
+		sidecarConfig := env.Get(gvk.Sidecar, changedSidecar.Name, changedSidecar.Namespace)
+		if sidecarConfig == nil {
+			// sidecar config is deleted remove it from the indexes
+			m := ps.sidecarIndex.sidecarsByNamespaceWithSelector[changedSidecar.Namespace]
+			delete(m, changedSidecar.Name)
+			if len(m) == 0 {
+				delete(ps.sidecarIndex.sidecarsByNamespaceWithSelector, changedSidecar.Namespace)
+			}
+
+			m = ps.sidecarIndex.sidecarsByNamespaceWithoutSelector[changedSidecar.Namespace]
+			delete(m, changedSidecar.Name)
+			if len(m) == 0 {
+				delete(ps.sidecarIndex.sidecarsByNamespaceWithoutSelector, changedSidecar.Namespace)
+			}
+
+			// the root namespace sidecar config is deleted
+			if rootNSConfig != nil && rootNSConfig.Namespace == changedSidecar.Namespace && rootNSConfig.Name == changedSidecar.Name {
+				rootNSConfig = nil
+			}
+
+			continue
+		}
+
+		// rebuild sidecar scope for this sidecar config.
+		// TODO: we could do better than this and only add/remove/update the actually changed resources.
+		sidecar := sidecarConfig.Spec.(*networking.Sidecar)
+		if sidecar.WorkloadSelector != nil {
+			m, f := ps.sidecarIndex.sidecarsByNamespaceWithSelector[changedSidecar.Namespace]
+			if !f {
+				m = make(map[string]*SidecarScope)
+				ps.sidecarIndex.sidecarsByNamespaceWithSelector[changedSidecar.Namespace] = m
+			}
+			m[changedSidecar.Name] = ConvertToSidecarScope(ps, sidecarConfig, changedSidecar.Namespace)
+		} else {
+			if sidecarConfig.Namespace == ps.Mesh.RootNamespace {
+				newRootNSConfig = sidecarConfig
+			}
+
+			m, f := ps.sidecarIndex.sidecarsByNamespaceWithoutSelector[changedSidecar.Namespace]
+			if !f {
+				m = make(map[string]*SidecarScope)
+				ps.sidecarIndex.sidecarsByNamespaceWithoutSelector[changedSidecar.Namespace] = m
+			}
+			m[changedSidecar.Name] = ConvertToSidecarScope(ps, sidecarConfig, changedSidecar.Namespace)
 		}
 	}
-	ps.sidecarIndex.meshRootSidecarConfig = rootNSConfig
+
+	if rootNSConfig == nil {
+		ps.sidecarIndex.meshRootSidecarConfig = newRootNSConfig
+	} else {
+		ps.sidecarIndex.meshRootSidecarConfig = rootNSConfig
+	}
+	// clear derived sidecars cache to allow newly created sidecar configs to be used
+	// TODO: we could probably be more intelligent about this, and only clear the derived
+	// sidecars that are affected by the changed sidecars or resources, but i'm not sure
+	// it's worth the trouble.
+	ps.sidecarIndex.meshRootSidecarsByNamespace = make(map[string]*SidecarScope)
+	ps.sidecarIndex.defaultSidecarsByNamespace = make(map[string]*SidecarScope)
 
 	return nil
 }
 
+// sidecarsUsingResources returns a set of sidecars that are affected by the given resources
+func sidecarsUsingResources(sidecarIndex sidecarIndex, services []ConfigKey, destinationRules []ConfigKey, virtualServices []ConfigKey) sets.Set[ConfigKey] {
```

Comment:
For service, this optimization is very limited, because when endpoint update, we mark it service config updated